File: examples/s3-support-adaptive-ranking/src/test-s3.ts
#!/usr/bin/env -S npx tsx
// Run: RATEL_S3_TEST_BUCKET=my-bucket ./src/test-s3.ts
import assert from "node:assert/strict";
import { IntentGraph, S3IntentGraphStorage, StaleIntentGraphError } from "@ratel-ai/sdk";

const rawBucket = process.env.RATEL_S3_TEST_BUCKET;
if (!rawBucket) {
  console.error("set RATEL_S3_TEST_BUCKET (and AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY)");
  process.exit(1);
}
const bucket: string = rawBucket;
const key = process.env.RATEL_S3_TEST_KEY ?? "ratel/intent-graph-s3-test.json";
const region = process.env.RATEL_S3_TEST_REGION ?? "us-east-1";
const endpoint = process.env.RATEL_S3_TEST_ENDPOINT;
const forcePathStyle =
  process.env.RATEL_S3_TEST_FORCE_PATH_STYLE === undefined
    ? undefined
    : process.env.RATEL_S3_TEST_FORCE_PATH_STYLE !== "false";
const idleTimeoutMs = process.env.RATEL_S3_TEST_IDLE_TIMEOUT_MS
  ? Number(process.env.RATEL_S3_TEST_IDLE_TIMEOUT_MS)
  : undefined;

const graph = (rev: number) =>
  IntentGraph.fromJson(JSON.stringify({ v: 1, built_from_ts: Date.now(), rev, intents: [] }));
const storage = () =>
  new S3IntentGraphStorage({ bucket, key, region, endpoint, forcePathStyle, idleTimeoutMs });

async function main() {
  const initial = storage();
  const before = await initial.load();
  const rev = (before?.rev ?? 0) + 1;
  await initial.save(graph(rev));

  const reloaded = await storage().load();
  assert.ok(reloaded, "object should exist after save");
  assert.equal(reloaded.rev, rev, "reloaded rev should match what was written");

  const skipCheck = storage();
  await skipCheck.load();
  await skipCheck.save(graph(rev)); // save-when-changed: no-op, must not throw

  const writerA = storage();
  const writerB = storage();
  await writerA.load();
  await writerB.load();
  await writerA.save(graph(rev + 1));
  await assert.rejects(
    () => writerB.save(graph(rev + 2)),
    StaleIntentGraphError,
    "stale writer should raise StaleIntentGraphError",
  );

  const final = await storage().load();
  assert.equal(final?.rev, rev + 1, "stale writer's save must not have landed");

  console.log(`PASS (s3-support-adaptive-ranking): s3://${bucket}/${key}, rev=${final.rev}`);
}

main().catch((err) => {
  console.error("FAIL (s3-support-adaptive-ranking):", err);
  process.exit(1);
});

File: examples/s3-support-adaptive-ranking-python/test_s3.py
#!/usr/bin/env -S uv run python3
"""Run: RATEL_S3_TEST_BUCKET=my-bucket ./test_s3.py"""

from __future__ import annotations

import asyncio
import json
import os
import sys
import time

from ratel_ai import S3IntentGraphStorage, IntentGraph, StaleIntentGraphError

bucket = os.environ.get("RATEL_S3_TEST_BUCKET")
if not bucket:
    print("set RATEL_S3_TEST_BUCKET (and AWS_ACCESS_KEY_ID / AWS_SECRET_ACCESS_KEY)", file=sys.stderr)
    sys.exit(1)

key = os.environ.get("RATEL_S3_TEST_KEY", "ratel/intent-graph-s3-test-v1.json")
region = os.environ.get("RATEL_S3_TEST_REGION", "us-east-1")
endpoint = os.environ.get("RATEL_S3_TEST_ENDPOINT")
_force_path_style_env = os.environ.get("RATEL_S3_TEST_FORCE_PATH_STYLE")
force_path_style = None if _force_path_style_env is None else _force_path_style_env != "false"


def graph(rev: int) -> IntentGraph:
    return IntentGraph.from_json(
        json.dumps({"v": 1, "built_from_ts": int(time.time() * 1000), "rev": rev, "intents": []})
    )


def storage() -> S3IntentGraphStorage:
    return S3IntentGraphStorage(
        bucket=bucket,
        key=key,
        region=region,
        endpoint=endpoint,
        force_path_style=force_path_style,
    )


async def main() -> None:
    initial = storage()
    before = await initial.load()
    rev = (before.rev if before else 0) + 1
    await initial.save(graph(rev))

    reloaded = await storage().load()
    assert reloaded is not None, "object should exist after save"
    assert reloaded.rev == rev, "reloaded rev should match what was written"

    skip_check = storage()
    await skip_check.load()
    await skip_check.save(graph(rev))  # save-when-changed: no-op, must not raise

    writer_a = storage()
    writer_b = storage()
    await writer_a.load()
    await writer_b.load()
    await writer_a.save(graph(rev + 1))
    try:
        await writer_b.save(graph(rev + 2))
        raise AssertionError("stale writer should raise StaleIntentGraphError")
    except StaleIntentGraphError:
        pass

    final = await storage().load()
    assert final is not None and final.rev == rev + 1, "stale writer's save must not have landed"

    print(f"PASS (s3-support-adaptive-ranking-python): s3://{bucket}/{key}, rev={final.rev}")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except Exception as err:  # noqa: BLE001
        print(f"FAIL (s3-support-adaptive-ranking-python): {err}", file=sys.stderr)
        sys.exit(1)
